### PR TITLE
eval-config.nix: Set modulePath to the path to NixOS modules

### PR DESCRIPTION
**Pull request by shlevy**
Alternative (not necessarily superior) to #8188

### CODE REVIEW

**Comment by copumpkin** [comment]
Fine by me! All I want is for it to work, so if this works, so be it :smile: 

Should this also remove extra-arguments.nix and the reference from the module list?

**Comment by domenkozar** [comment]
Also see #7835

**Comment by copumpkin** [comment]
That one seems far more complicated and scary, to my uneducated eye.

File: nixos/lib/eval-config.nix [hunk 1]
```diff
@@ -22,8 +22,17 @@
...
 , prefix ? []
 , lib ? import ../../lib
-}:
+}@args: if !(builtins ? modulePath) then let
+  scope = {
+    scopedImport = newScope: scopedImport (scope // newScope);
 
+    import = scopedImport scope;
+
+    modulePath = ../modules;
```
Comment (copumpkin) [inline nixos/lib/eval-config.nix]: Should it be `modulesPath`?

**Comment by copumpkin** [comment]
Can you explain briefly what the `scopedImport` is changing functionality-wise?

File: nixos/lib/eval-config.nix [hunk 2]
```diff
@@ -22,8 +22,17 @@
   check ? true
 , prefix ? []
 , lib ? import ../../lib
-}:
+}@args: if !(builtins ? modulePath) then let
```
Comment (copumpkin) [inline nixos/lib/eval-config.nix]: Huh?

**Comment by copumpkin** [comment]
I think I'll go ahead and merge (a modified version of) this, given the limited time and difficulty contacting everyone involved. If #7835 ends up being a better solution (I don't understand it well enough to judge, but it looks more complicated and like @edolstra has more objections) we can undo this and put that in.

**Comment by copumpkin** [comment]
Looking up [`scopedImport`](https://github.com/NixOS/nix/commit/c273c15cb13bb86420dda1e5341a4e19517532b5) it sort of scares me a bit. I guess it allows importing expressions that contain free variables. Your proposed change would allow me to refer to `modulesPath` without even declaring it as a function parameter in my configuration.nix, right? Scares me a little, but if that's the direction we're heading (and why we added `scopedImport` in the first place), I don't have a problem with it.

**Comment by domenkozar** [comment]
I'd wait for a comment by @nbp

**Comment by copumpkin** [comment]
@shlevy actually, I'm trying this out and it doesn't seem to help. Passing more `extraArgs` in eval-config.nix just tells lib/modules.nix to stick it into `_module.args`, which then fails in the exact same way :cry:

**Comment by shlevy** [comment]
Hey, sorry! I fired this off before leaving for a family trip, I didn't mean to hold up your work just put this out there as a possibility! The idea would be that instead of taking `modulePath` as an argument to your module, you can just refer to `modulePath` (or `builtins.modulePath`) in each module. This bypasses module arguments all together.

File: nixos/lib/eval-config.nix (hunk 1 above)
Comment (shlevy) [inline nixos/lib/eval-config.nix]: Yeah

File: nixos/lib/eval-config.nix (hunk 2 above)
Comment (shlevy) [inline nixos/lib/eval-config.nix]: Not on the initial import, but on the scoped reimport there will be! This avoids an infinte loop of reimports.

Comment (copumpkin) [inline nixos/lib/eval-config.nix]: Yeah, I see now, thanks :smile:

**Comment by shlevy** [comment]
Closed in favor of #8204

**Pull request closed without merging**